[MM-49179] Global Drafts Onboarding Tour Point

mylonsuren opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

Adds a tour point for Global Drafts that appears once the user creates their first draft.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

https://mattermost.atlassian.net/browse/MM-49179

<!--
#### Related Pull Requests

List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.

- Has server changes (please link here)
- Has mobile changes (please link here)
-->
#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

<img width="2990" alt="image" src="https://user-images.githubusercontent.com/23694620/208785420-97babbcf-2375-46be-ae0b-dead5654cdcc.png">


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
NONE
```


[MM-49179]: https://mattermost.atlassian.net/browse/MM-49179?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

matthewbirtch commented:
@jgilliam17 could you help with the feature flag on this test server?

jgilliam17 commented:
@matthewbirtch Global Drafts feature flag is on

jgilliam17 commented:
@mylonsuren Onboarding tour is triggered after first character is typed, maybe we should add a bit of delay.

matthewbirtch commented:
@mylonsuren the UI looks good, but I'm wondering about the logic.
- I thought the Drafts item didn't show in the LSH until the user lost focus on message input. However, it seems we show the Drafts item as soon as you start typing. We should look at change this I think.
- Rather than waiting until I've created a draft, somehow I got the tour point as soon as I logged in to the test server with the sysadmin account.

For testing purposes, is there any way for me to clear the status of the tour point? Or do I need to create a new account to see it again?

mylonsuren commented:
@matthewbirtch

> I thought the Drafts item didn't show in the LHS until the user lost focus on message input. However, it seems we show the Drafts item as soon as you start typing. We should look at change this I think. Also, it seems in the RHS we aren't showing the LHS Drafts item until losing focus on the input. We just need to be consistent with behaviour in the main channel input

Yes you're right, currently the LHS only shows the Drafts item once the user unfocused the message input. I added a change here to make it appear once the draft is saved locally, and have reverted it now. The associated change for the RHS was added in this PR (https://github.com/mattermost/mattermost-webapp/pull/11803), and have also reverted it there as it is not associated with the main fix there. If anything that can be discussed separately and raised as another ticket if necessary. 

> Rather than waiting until I've created a draft, somehow I got the tour point as soon as I logged in to the test server with the sysadmin account.

Let me look into this as it should only occur once the Drafts item on the LHS becomes visible for the first time. Did you have any drafts created in any of the channels prior to the feature flag being enabled?

> For testing purposes, is there any way for me to clear the status of the tour point? Or do I need to create a new account to see it again?

I could manually override the logic that prevents the tour point showing again and then revert prior to Dev / QA Review. In prior PRs how have tour points generally been tested? @jgilliam17 is there a way to "refresh" the account in the cloud server so that tour points are reset?

jgilliam17 commented:
>is there a way to "refresh" the account in the cloud server so that tour points are reset?

I am not sure. What I've done in the past is create new user via invite link

matthewbirtch commented:
> Did you have any drafts created in any of the channels prior to the feature flag being enabled?

@mylonsuren I'm not sure, but that sounds like what happened here. Probably a non-issue

matthewbirtch commented:
> I could manually override the logic that prevents the tour point showing again and then revert prior to Dev / QA Review. In prior PRs how have tour points generally been tested? @jgilliam17 is there a way to "refresh" the account in the cloud server so that tour points are reset?

It's ok, it's easy enough to test with new users. Let's now worry about it

matthewbirtch reviewed (approved):
Looking great to me! Awesome work. All looks good now to me

AshishDhama commented on components/drafts/drafts_link/drafts_link.tsx:
```diff
@@ -0,0 +1,79 @@
+                    role='listitem'
+                    tabIndex={0}
+                >
+                    <i
```
Use compass icons

mylonsuren replied:
Done 👍

AshishDhama commented on components/drafts/drafts_link/drafts_tour_tip/drafts_tour_tip.tsx:
```diff
@@ -0,0 +1,119 @@
+                    step={1}
+                    singleTip={true}
+                    showOptOut={false}
+                    interactivePunchOut={true}
```
I think interactive isn't working as expected, so better would be to make `interactivePunchOut={false}`. A contributor is working on it. Otherwise, I will fix it soon.

mylonsuren replied:
I've set it to false for now 👍

AshishDhama reviewed (approved):
LGTM!

koox00 commented on components/drafts/drafts_link/drafts_link.tsx:
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Did you meant to create this file, again?

mylonsuren replied:
Yes sorry, good catch. It looks like the deletion of the original `drafts_link` files didn't get tracked. I've removed those extra files now 👍

koox00 commented on components/drafts/drafts_link/drafts_tour_tip/drafts_tour_tip.tsx:
```diff
@@ -0,0 +1,119 @@
+import {setDraftsTourTipPreference} from 'actions/views/drafts';
+
+const title = (
+    <span className='title-with-new-badge'>
```
FYI you could avoid using a new class by using utility classes we have (boostrap) 0/5

```suggestion
    <span className='d-flex align-items-center'>
```

mylonsuren replied:
Good idea 👍

mattermod commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

mylonsuren commented:
/update-branch

jgilliam17 commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_GLOBALDRAFTS=true"

mattermod commented:
Triggering E2E testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_GLOBALDRAFTS=true]```

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/284645) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/284645)

jgilliam17 commented:
Thanks @mylonsuren 
Some tests are failing due to global drafts tour
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/284645-2f36eda-onprem-ent-webapp-pr-11846-mm-ee-test:2f36eda/mochawesome.html)

mattermost-build commented:
Triggering E2E testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_GLOBALDRAFTS=true]```

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/286450) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/286450)

mylonsuren commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_GLOBALDRAFTS=true"

mattermost-build commented:
Triggering E2E testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_GLOBALDRAFTS=true]```

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/286491) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/286491)

mylonsuren commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_GLOBALDRAFTS=true"

mattermost-build commented:
Triggering E2E testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_GLOBALDRAFTS=true]```

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/286515) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/286515)

mylonsuren commented:
@jgilliam17 I've made some changes to address the tour tip blocking other e2e tests. Could you please review and let me know if there are still any persisting issues. Thanks 😄

jgilliam17 commented:
Thanks @mylonsuren 
I am checking the report right now, will post shortly

jgilliam17 commented:
Thanks @mylonsuren and apologies for delay
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/286798-bcf5b19-onprem-ent-webapp-pr-11846-mm-ee-test:bcf5b19/mochawesome.html) looks good, no PR related failures.
